lib/Sema/CSFix.cpp

```diff
@@ -871,6 +871,47 @@ IgnoreContextualType *IgnoreContextualType::create(ConstraintSystem &cs,
       IgnoreContextualType(cs, resultTy, specifiedTy, locator);
 }
 
+bool RemoveUnnecessaryCoercion::diagnose(bool asNote) const {
+  auto &cs = getConstraintSystem();
+  UnnecessaryCoercionFailure failure(cs, getFromType(), getToType(),
+                                     getLocator());
+  return failure.diagnose(asNote);
+}
+
+bool RemoveUnnecessaryCoercion::attempt(ConstraintSystem &cs, Type fromType,
+                                        Type toType,
+                                        ConstraintLocatorBuilder locator) {
+  auto last = locator.last();
+  bool isExplicitCoercion =
+      last && last->is<LocatorPathElt::ExplicitTypeCoercion>();
+  if (!isExplicitCoercion)
+    return false;
+
+  auto expr = cast<CoerceExpr>(locator.getAnchor());
+  auto toTypeRepr = expr->getCastTypeLoc().getTypeRepr();
+
+  // Only diagnosing for coercion where the left-side is a DeclRefExpr
+  // or a explicit/implicit coercion e.g. Double(1) as Double
+  if (!isa<ImplicitlyUnwrappedOptionalTypeRepr>(toTypeRepr) &&
+      (isa<DeclRefExpr>(expr->getSubExpr()) ||
+       isa<CoerceExpr>(expr->getSubExpr()))) {
```

Comment:
The question is - if we are going to keep track of where the type came from do we really need to check whether sub-expression is a declaration or another coercion? It seems like source information of the binding should be enough do distinguish that?

Reply:
Yeah, that would also allow this to diagnose cases like 
```swift
_ = "Str" as String // Since inferred string type from the literal is not coming from the coercion
_ = 1 as Int // Since inferred string type from the literal is not coming from the coercion
```
So it will be a good thing :) 
Maybe in this context, we should have a more tailored diagnostic message mentioning
the type of subExpr was inferred by context.